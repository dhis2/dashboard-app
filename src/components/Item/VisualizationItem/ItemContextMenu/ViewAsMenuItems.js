import i18n from '@dhis2/d2-i18n'
import { IconVisualizationColumn16, IconTable16, IconWorld16 } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React from 'react'
import {
    CHART,
    MAP,
    REPORT_TABLE,
    EVENT_CHART,
    EVENT_REPORT,
    EVENT_VISUALIZATION,
    isTrackerDomainType,
    hasMapView,
} from '../../../../modules/itemTypes.js'
import MenuItem from '../../../MenuItemWithTooltip.js'
import getThematicMapViews from '../getThematicMapViews.js'

const ViewAsMenuItems = ({
    type,
    activeType,
    visualization,
    onActiveTypeChanged,
}) => {
    const isTrackerType = isTrackerDomainType(type)

    const onViewTable = () =>
        onActiveTypeChanged(isTrackerType ? EVENT_REPORT : REPORT_TABLE)

    const onViewChart = () =>
        onActiveTypeChanged(isTrackerType ? EVENT_CHART : CHART)

    const onViewMap = () => onActiveTypeChanged(MAP)

    const notSupported = type === MAP && !getThematicMapViews(visualization)

    return (
        <>
            {![CHART, EVENT_CHART].includes(activeType) && (
                <MenuItem
                    tooltip={
                        notSupported
                            ? i18n.t("This map can't be displayed as a chart")
                            : null
                    }
                    label={i18n.t('View as Chart')}
                    onClick={onViewChart}
                    disabled={notSupported}
                    icon={<IconVisualizationColumn16 />}
                />
            )}
            {![REPORT_TABLE, EVENT_REPORT, EVENT_VISUALIZATION].includes(
                activeType
            ) && (
                <MenuItem
                    tooltip={
                        notSupported
                            ? i18n.t("This map can't be displayed as a table")
                            : null
                    }
                    label={i18n.t('View as Table')}
                    onClick={onViewTable}
                    disabled={notSupported}
                    icon={<IconTable16 />}
                />
            )}
            {hasMapView(type) && activeType !== MAP && (
                <MenuItem
                    label={i18n.t('View as Map')}
                    onClick={onViewMap}
                    icon={<IconWorld16 />}
                />
            )}
        </>
    )
}

ViewAsMenuItems.propTypes = {
    activeType: PropTypes.string,
    type: PropTypes.string,
    visualization: PropTypes.object,
    onActiveTypeChanged: PropTypes.func,
}

export default ViewAsMenuItems
