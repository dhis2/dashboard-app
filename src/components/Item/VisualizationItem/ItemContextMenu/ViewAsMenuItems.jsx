import { VIS_TYPE_OUTLIER_TABLE } from '@dhis2/analytics'
import i18n from '@dhis2/d2-i18n'
import { IconVisualizationColumn16, IconTable16, IconWorld16 } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React from 'react'
import {
    CHART,
    MAP,
    REPORT_TABLE,
    EVENT_CHART,
    EVENT_REPORT,
    EVENT_VISUALIZATION,
    isTrackerDomainType,
    hasMapView,
} from '../../../../modules/itemTypes.js'
import MenuItem from '../../../MenuItemWithTooltip.jsx'
import getThematicMapViews from '../getThematicMapViews.js'

const ViewAsMenuItems = ({
    type,
    activeType,
    visualization,
    onActiveTypeChanged,
}) => {
    const isTrackerType = isTrackerDomainType(type)

    const onViewTable = () =>
        onActiveTypeChanged(isTrackerType ? EVENT_REPORT : REPORT_TABLE)

    const onViewChart = () =>
        onActiveTypeChanged(isTrackerType ? EVENT_CHART : CHART)

    const onViewMap = () => onActiveTypeChanged(MAP)

    const notSupported =
        (type === MAP && !getThematicMapViews(visualization)) ||
        (type === CHART && visualization.type === VIS_TYPE_OUTLIER_TABLE)

    const getNotSupportedMessage = (viewAs) => {
        if (type === MAP && !getThematicMapViews(visualization)) {
            return viewAs === 'chart'
                ? i18n.t("This map can't be displayed as a chart")
                : i18n.t("This map can't be displayed as a pivot table")
        }

        if (type === CHART && visualization.type === VIS_TYPE_OUTLIER_TABLE) {
            return viewAs === 'table'
                ? i18n.t(
                      "This visualization can't be displayed as a pivot table"
                  )
                : i18n.t("This visualization can't be displayed as a map")
        }

        return null
    }

    return (
        <>
            {![CHART, EVENT_CHART].includes(activeType) && (
                <MenuItem
                    tooltip={getNotSupportedMessage('chart')}
                    label={i18n.t('View as Chart')}
                    onClick={onViewChart}
                    disabled={notSupported}
                    icon={<IconVisualizationColumn16 />}
                />
            )}
            {![REPORT_TABLE, EVENT_REPORT, EVENT_VISUALIZATION].includes(
                activeType
            ) && (
                <MenuItem
                    tooltip={getNotSupportedMessage('table')}
                    label={i18n.t('View as Pivot table')}
                    onClick={onViewTable}
                    disabled={notSupported}
                    icon={<IconTable16 />}
                />
            )}
            {hasMapView(type) && activeType !== MAP && (
                <MenuItem
                    tooltip={getNotSupportedMessage('map')}
                    label={i18n.t('View as Map')}
                    onClick={onViewMap}
                    disabled={notSupported}
                    icon={<IconWorld16 />}
                />
            )}
        </>
    )
}

ViewAsMenuItems.propTypes = {
    activeType: PropTypes.string,
    type: PropTypes.string,
    visualization: PropTypes.object,
    onActiveTypeChanged: PropTypes.func,
}

export default ViewAsMenuItems
