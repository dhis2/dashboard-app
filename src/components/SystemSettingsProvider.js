import { useDataEngine } from '@dhis2/app-runtime'
import PropTypes from 'prop-types'
import React, { useContext, useState, useEffect, createContext } from 'react'
import {
    systemSettingsQuery,
    renameSystemSettings,
    DEFAULT_SETTINGS,
} from '../api/systemSettings'

export const SystemSettingsCtx = createContext({})

const SystemSettingsProvider = ({ children }) => {
    const [settings, setSettings] = useState([])
    const engine = useDataEngine()

    useEffect(() => {
        async function fetchData() {
            const { systemSettings } = await engine.query({
                systemSettings: systemSettingsQuery,
            })

            setSettings(
                Object.assign(
                    {},
                    renameSystemSettings(DEFAULT_SETTINGS),
                    renameSystemSettings(systemSettings)
                )
            )
        }
        fetchData()
    }, [])

    return (
        <SystemSettingsCtx.Provider
            value={{
                settings,
            }}
        >
            {children}
        </SystemSettingsCtx.Provider>
    )
}

SystemSettingsProvider.propTypes = {
    children: PropTypes.node,
}

export default SystemSettingsProvider

export const useSystemSettings = () => useContext(SystemSettingsCtx)
