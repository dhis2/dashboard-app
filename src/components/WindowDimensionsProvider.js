import React, { createContext, useContext, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import debounce from 'lodash/debounce'

export const WindowDimensionsCtx = createContext(null)

const windowDims = () => ({
    height: window.innerHeight,
    width: window.innerWidth,
})

const WindowDimensionsProvider = ({ children }) => {
    const [dimensions, setDimensions] = useState(windowDims())

    useEffect(() => {
        const handleResize = debounce(() => {
            setDimensions(windowDims())
        }, 200)
        window.addEventListener('resize', handleResize)
        return () => {
            window.removeEventListener('resize', handleResize)
        }
    }, [])

    return (
        <WindowDimensionsCtx.Provider value={dimensions}>
            {children}
        </WindowDimensionsCtx.Provider>
    )
}

WindowDimensionsProvider.propTypes = {
    children: PropTypes.node.isRequired,
}

export default WindowDimensionsProvider

export const useWindowDimensions = () => {
    return useContext(WindowDimensionsCtx)
}
