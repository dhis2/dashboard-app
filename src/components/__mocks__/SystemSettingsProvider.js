import PropTypes from 'prop-types'
import React, { useContext, useState, createContext } from 'react'

export const SystemSettingsCtx = createContext({})

const SystemSettingsProvider = ({ children }) => {
    const [settings] = useState({
        keyDashboardContextMenuItemOpenInRelevantApp: true,
        keyDashboardContextMenuItemShowInterpretationsAndDetails: true,
        keyDashboardContextMenuItemSwitchViewType: true,
        keyDashboardContextMenuItemViewFullscreen: true,
        keyGatherAnalyticalObjectStatisticsInDashboardViews: false,
    })

    return (
        <SystemSettingsCtx.Provider
            value={{
                settings,
            }}
        >
            {children}
        </SystemSettingsCtx.Provider>
    )
}

SystemSettingsProvider.propTypes = {
    children: PropTypes.node,
}

export default SystemSettingsProvider

export const useSystemSettings = () => useContext(SystemSettingsCtx)
