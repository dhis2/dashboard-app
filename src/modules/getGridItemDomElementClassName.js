export const getGridItemDomElementClassName = itemId =>
    `reactgriditem-${itemId}`
