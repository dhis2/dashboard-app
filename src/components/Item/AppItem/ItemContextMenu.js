import i18n from '@dhis2/d2-i18n'
import {
    colors,
    Button,
    Menu,
    Popover,
    IconFullscreen16,
    IconFullscreenExit16,
    IconLaunch16,
    IconMore24,
} from '@dhis2/ui'
import PropTypes from 'prop-types'
import React, { useState, createRef } from 'react'
import { isSmallScreen } from '../../../modules/smallScreen.js'
import MenuItem from '../../MenuItemWithTooltip.js'
import { useSystemSettings } from '../../SystemSettingsProvider.js'
import { useWindowDimensions } from '../../WindowDimensionsProvider.js'
import { isElementFullscreen } from '../fullscreenUtil.js'

export const ItemContextMenu = ({
    appName,
    appUrl,
    item,
    fullscreenSupported,
    onToggleFullscreen,
    loadItemFailed,
}) => {
    const [menuIsOpen, setMenuIsOpen] = useState(false)
    const { width } = useWindowDimensions()

    const { allowVisOpenInApp, allowVisFullscreen } =
        useSystemSettings().systemSettings

    const fullscreenAllowed = fullscreenSupported && allowVisFullscreen

    const noOptionsEnabled = !allowVisOpenInApp && !fullscreenAllowed

    if (noOptionsEnabled || (!allowVisOpenInApp && loadItemFailed)) {
        return null
    }

    const toggleFullscreen = () => {
        onToggleFullscreen()
        closeMenu()
    }

    const openMenu = () => {
        setMenuIsOpen(true)
    }
    const closeMenu = () => setMenuIsOpen(false)

    console.log('AppItem', item)

    const buttonRef = createRef()

    return isElementFullscreen(item.id) ? (
        <Button small secondary onClick={onToggleFullscreen}>
            <span data-testid="exit-fullscreen-button">
                <IconFullscreenExit16 color={colors.grey600} />
            </span>
        </Button>
    ) : (
        <>
            <div ref={buttonRef}>
                <Button
                    small
                    secondary
                    onClick={openMenu}
                    dataTest="dashboarditem-menu-button"
                    icon={<IconMore24 color={colors.grey700} />}
                />
            </div>
            {menuIsOpen && (
                <Popover
                    reference={buttonRef}
                    placement="auto-start"
                    arrow={false}
                    onClickOutside={closeMenu}
                >
                    <Menu dense>
                        {allowVisOpenInApp && !isSmallScreen(width) && (
                            <MenuItem
                                icon={<IconLaunch16 />}
                                disabledWhenOffline={false}
                                label={i18n.t('Open in {{appName}}', {
                                    appName,
                                })}
                                href={appUrl}
                                target="_blank"
                            />
                        )}
                        {fullscreenAllowed && !loadItemFailed && (
                            <MenuItem
                                disabledWhenOffline={false}
                                icon={<IconFullscreen16 />}
                                label={i18n.t('View fullscreen')}
                                onClick={toggleFullscreen}
                            />
                        )}
                    </Menu>
                </Popover>
            )}
        </>
    )
}

ItemContextMenu.propTypes = {
    appName: PropTypes.string,
    appUrl: PropTypes.string,
    fullscreenSupported: PropTypes.bool,
    item: PropTypes.object,
    loadItemFailed: PropTypes.bool,
    onToggleFullscreen: PropTypes.func,
}
