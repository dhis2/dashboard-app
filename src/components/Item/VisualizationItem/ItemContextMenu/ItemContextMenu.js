import React, { useState, createRef } from 'react'
import PropTypes from 'prop-types'
import { useConfig } from '@dhis2/app-runtime'
import {
    isSingleValue,
    isYearOverYear,
    VIS_TYPE_GAUGE,
    VIS_TYPE_PIE,
} from '@dhis2/analytics'
import { Button, Menu, Popover, MenuItem, Divider } from '@dhis2/ui'
import i18n from '@dhis2/d2-i18n'
import LaunchIcon from '@material-ui/icons/Launch'
import ViewAsMenuItems from './ViewAsMenuItems'
import { useWindowDimensions } from '../../../WindowDimensionsProvider'
import { isSmallScreen } from '../../../../modules/smallScreen'

import {
    ThreeDots,
    SpeechBubble,
    Fullscreen,
    ExitFullscreen,
} from '../assets/icons'
import { getLink } from '../Visualization/plugin'
import { getAppName } from '../../../../modules/itemTypes'
import { useSystemSettings } from '../../../SystemSettingsProvider'

const ItemContextMenu = props => {
    const [menuIsOpen, setMenuIsOpen] = useState(false)
    const { width } = useWindowDimensions()
    const { baseUrl } = useConfig()

    const {
        allowVisOpenInApp,
        allowVisShowInterpretations,
        allowVisViewAs,
        allowVisFullscreen,
    } = useSystemSettings().settings

    const fullscreenAllowed = props.fullscreenSupported && allowVisFullscreen

    const noOptionsEnabled =
        !allowVisOpenInApp &&
        !allowVisShowInterpretations &&
        !allowVisViewAs &&
        !fullscreenAllowed

    if (noOptionsEnabled || (!allowVisOpenInApp && props.loadItemFailed)) {
        return null
    }

    const toggleInterpretations = () => {
        props.onToggleFooter()
        if (menuIsOpen) {
            closeMenu()
        }
    }

    const toggleFullscreen = () => {
        props.onToggleFullscreen()
        closeMenu()
    }

    const onActiveTypeChanged = type => {
        closeMenu()
        props.onSelectActiveType(type)
    }

    const openMenu = () => {
        setMenuIsOpen(true)
    }
    const closeMenu = () => setMenuIsOpen(false)

    const { item, visualization, loadItemFailed, activeType } = props
    const type = visualization.type || item.type
    const canViewAs =
        allowVisViewAs &&
        !isSingleValue(type) &&
        !isYearOverYear(type) &&
        type !== VIS_TYPE_GAUGE &&
        type !== VIS_TYPE_PIE

    const interpretationMenuLabel = props.activeFooter
        ? i18n.t(`Hide interpretations and details`)
        : i18n.t(`Show interpretations and details`)

    const buttonRef = createRef()

    return props.isFullscreen ? (
        <Button small secondary onClick={props.onToggleFullscreen}>
            <ExitFullscreen />
        </Button>
    ) : (
        <>
            <div ref={buttonRef}>
                <Button
                    small
                    secondary
                    onClick={openMenu}
                    dataTest="dashboarditem-menu-button"
                >
                    <ThreeDots />
                </Button>
            </div>
            {menuIsOpen && (
                <Popover
                    reference={buttonRef}
                    placement="auto-start"
                    arrow={false}
                    onClickOutside={closeMenu}
                >
                    <Menu>
                        {canViewAs && !loadItemFailed && (
                            <>
                                <ViewAsMenuItems
                                    type={item.type}
                                    activeType={activeType}
                                    onActiveTypeChanged={onActiveTypeChanged}
                                />
                                {(allowVisShowInterpretations ||
                                    (allowVisOpenInApp &&
                                        !isSmallScreen(width)) ||
                                    fullscreenAllowed) && (
                                    <span data-testid="divider">
                                        <Divider />
                                    </span>
                                )}
                            </>
                        )}
                        {allowVisOpenInApp && !isSmallScreen(width) && (
                            <MenuItem
                                dense
                                icon={
                                    <LaunchIcon style={{ fill: '#6e7a8a' }} />
                                }
                                label={i18n.t('Open in {{appName}} app', {
                                    appName: getAppName(item.type),
                                })}
                                href={getLink(item, baseUrl)}
                                target="_blank"
                            />
                        )}
                        {allowVisShowInterpretations && !loadItemFailed && (
                            <MenuItem
                                dense
                                icon={<SpeechBubble />}
                                label={interpretationMenuLabel}
                                onClick={toggleInterpretations}
                            />
                        )}
                        {fullscreenAllowed && !loadItemFailed && (
                            <MenuItem
                                dense
                                icon={<Fullscreen />}
                                label={i18n.t('View fullscreen')}
                                onClick={toggleFullscreen}
                            />
                        )}
                    </Menu>
                </Popover>
            )}
        </>
    )
}

ItemContextMenu.propTypes = {
    activeFooter: PropTypes.bool,
    activeType: PropTypes.string,
    fullscreenSupported: PropTypes.bool,
    isFullscreen: PropTypes.bool,
    item: PropTypes.object,
    loadItemFailed: PropTypes.bool,
    visualization: PropTypes.object,
    onSelectActiveType: PropTypes.func,
    onToggleFooter: PropTypes.func,
    onToggleFullscreen: PropTypes.func,
}

export default ItemContextMenu
