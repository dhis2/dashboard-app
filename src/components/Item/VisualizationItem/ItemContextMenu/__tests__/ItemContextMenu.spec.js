import { fireEvent } from '@testing-library/dom'
import { render, waitFor, screen } from '@testing-library/react'
import React from 'react'
import { getGridItemDomElementClassName } from '../../../../../modules/getGridItemDomElementClassName'
import { useSystemSettings } from '../../../../SystemSettingsProvider'
import WindowDimensionsProvider from '../../../../WindowDimensionsProvider'
import ItemContextMenu from '../ItemContextMenu'

jest.mock('../../../../SystemSettingsProvider', () => ({
    useSystemSettings: jest.fn(),
}))

const mockSystemSettingsDefault = {
    settings: {
        allowVisOpenInApp: true,
        allowVisShowInterpretations: true,
        allowVisViewAs: true,
        allowVisFullscreen: true,
    },
}

const defaultProps = {
    item: {
        type: 'CHART',
        id: 'rainbowdash',
    },
    visualization: {
        type: 'BAR',
    },
    onSelectActiveType: Function.prototype,
    onToggleFooter: Function.prototype,
    activeFooter: false,
    activeType: 'CHART',
    fullscreenSupported: true,
    loadItemFailed: false,
}

test('renders just the button when menu closed', () => {
    useSystemSettings.mockImplementationOnce(() => mockSystemSettingsDefault)

    const { container } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...defaultProps} />
        </WindowDimensionsProvider>
    )

    expect(container).toMatchSnapshot()
})

test('renders exit fullscreen button', () => {
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
    const gridItemClassName = getGridItemDomElementClassName(
        defaultProps.item.id
    )

    const { rerender } = render(
        <WindowDimensionsProvider>
            <div className={gridItemClassName}>
                <ItemContextMenu {...defaultProps} />
            </div>
        </WindowDimensionsProvider>
    )

    document.fullscreenElement = document.querySelector(`.${gridItemClassName}`)

    rerender(
        <WindowDimensionsProvider>
            <div className={{ gridItemClassName }}>
                <ItemContextMenu {...defaultProps} />
            </div>
        </WindowDimensionsProvider>
    )

    document.fullscreenElement = null
    expect(screen.getByTestId('exit-fullscreen-button')).toBeTruthy()
})

test('renders popover menu for BAR chart', async () => {
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
    const props = Object.assign({}, defaultProps, {
        visualization: {
            type: 'BAR',
        },
    })

    const { getByRole, queryByText, queryByTestId } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...props} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View as Map')).toBeTruthy()
        expect(queryByText('View as Chart')).toBeNull()
        expect(queryByText('View as Table')).toBeTruthy()
        expect(queryByTestId('divider')).toBeTruthy()
        expect(queryByText('Open in Data Visualizer app')).toBeTruthy()
        expect(queryByText('Show details and interpretations')).toBeTruthy()
        expect(queryByText('View fullscreen')).toBeTruthy()
    })
})

test('renders popover menu for SINGLE_VALUE chart', async () => {
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
    const props = Object.assign({}, defaultProps, {
        visualization: {
            type: 'SINGLE_VALUE',
        },
    })

    const { getByRole, queryByTestId, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...props} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View as Map')).toBeNull()
        expect(queryByText('View as Chart')).toBeNull()
        expect(queryByText('View as Table')).toBeNull()
        expect(queryByTestId('divider')).toBeNull()
        expect(queryByText('Open in Data Visualizer app')).toBeTruthy()
        expect(queryByText('Show details and interpretations')).toBeTruthy()
        expect(queryByText('View fullscreen')).toBeTruthy()
    })
})

test('renders popover menu for YEAR_OVER_YEAR_LINE chart', async () => {
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
    const props = Object.assign({}, defaultProps, {
        visualization: {
            type: 'YEAR_OVER_YEAR_LINE',
        },
    })

    const { getByRole, queryByTestId, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...props} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View as Map')).toBeNull()
        expect(queryByText('View as Chart')).toBeNull()
        expect(queryByText('View as Table')).toBeNull()
        expect(queryByTestId('divider')).toBeNull()
        expect(queryByText('Open in Data Visualizer app')).toBeTruthy()
        expect(queryByText('Show details and interpretations')).toBeTruthy()
        expect(queryByText('View fullscreen')).toBeTruthy()
    })
})

test('renders popover menu for GAUGE chart', async () => {
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
    const props = Object.assign({}, defaultProps, {
        visualization: {
            type: 'GAUGE',
        },
    })

    const { getByRole, queryByTestId, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...props} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View as Map')).toBeNull()
        expect(queryByText('View as Chart')).toBeNull()
        expect(queryByText('View as Table')).toBeNull()
        expect(queryByTestId('divider')).toBeNull()
        expect(queryByText('Open in Data Visualizer app')).toBeTruthy()
        expect(queryByText('Show details and interpretations')).toBeTruthy()
        expect(queryByText('View fullscreen')).toBeTruthy()
    })
})

test('renders popover menu for PIE chart', async () => {
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
    const props = Object.assign({}, defaultProps, {
        visualization: {
            type: 'PIE',
        },
    })

    const { getByRole, queryByTestId, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...props} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View as Map')).toBeNull()
        expect(queryByText('View as Chart')).toBeNull()
        expect(queryByText('View as Table')).toBeNull()
        expect(queryByTestId('divider')).toBeNull()
        expect(queryByText('Open in Data Visualizer app')).toBeTruthy()
        expect(queryByText('Show details and interpretations')).toBeTruthy()
        expect(queryByText('View fullscreen')).toBeTruthy()
    })
})

test('renders popover menu for PIVOT_TABLE', async () => {
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
    const props = Object.assign({}, defaultProps, {
        item: {
            type: 'REPORT_TABLE',
        },
        visualization: {
            type: 'PIVOT_TABLE',
        },
        activeType: 'REPORT_TABLE',
    })

    const { getByRole, queryByTestId, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...props} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View as Map')).toBeTruthy()
        expect(queryByText('View as Chart')).toBeTruthy()
        expect(queryByText('View as Table')).toBeNull()
        expect(queryByTestId('divider')).toBeTruthy()
        expect(queryByText('Open in Data Visualizer app')).toBeTruthy()
        expect(queryByText('Show details and interpretations')).toBeTruthy()
        expect(queryByText('View fullscreen')).toBeTruthy()
    })
})

test('renders popover menu for MAP', async () => {
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
    const props = Object.assign({}, defaultProps, {
        item: {
            type: 'MAP',
        },
        visualization: {},
        activeType: 'MAP',
    })

    const { getByRole, queryByTestId, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...props} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View as Map')).toBeNull()
        expect(queryByText('View as Chart')).toBeTruthy()
        expect(queryByText('View as Table')).toBeTruthy()
        expect(queryByTestId('divider')).toBeTruthy()
        expect(queryByText('Open in Maps app')).toBeTruthy()
        expect(queryByText('Show details and interpretations')).toBeTruthy()
        expect(queryByText('View fullscreen')).toBeTruthy()
    })
})

test('renders popover menu when interpretations displayed', async () => {
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
    const props = Object.assign({}, defaultProps, {
        visualization: {
            type: 'BAR',
        },
        activeFooter: true,
    })

    const { getByRole, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...props} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('Hide details and interpretations')).toBeTruthy()
    })
})

test('does not render "View as" options if settings do not allow', async () => {
    useSystemSettings.mockImplementation(() => ({
        settings: Object.assign({}, mockSystemSettingsDefault.settings, {
            allowVisViewAs: false,
        }),
    }))

    const { getByRole, queryAllByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...defaultProps} />
        </WindowDimensionsProvider>
    )
    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryAllByText(/View as/i)).toHaveLength(0)
    })
})

test('does not render "Open in [app]" option if settings do not allow', async () => {
    useSystemSettings.mockImplementation(() => ({
        settings: Object.assign({}, mockSystemSettingsDefault.settings, {
            allowVisOpenInApp: false,
        }),
    }))

    const { getByRole, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...defaultProps} />
        </WindowDimensionsProvider>
    )
    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText(/Open in/i)).toBeNull()
    })
})

test('renders only View in App when item load failed', async () => {
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
    const props = Object.assign({}, defaultProps, {
        item: {
            type: 'MAP',
        },
        visualization: {},
        activeType: 'MAP',
        loadItemFailed: true,
    })

    const { getByRole, queryByTestId, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...props} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View as Map')).toBeNull()
        expect(queryByText('View as Chart')).toBeNull()
        expect(queryByText('View as Table')).toBeNull()
        expect(queryByTestId('divider')).toBeNull()
        expect(queryByText('Open in Maps app')).toBeTruthy()
        expect(queryByText('Show details and interpretations')).toBeNull()
        expect(queryByText('View fullscreen')).toBeNull()
    })
})

test('does not render "fullscreen" option if settings do not allow', async () => {
    useSystemSettings.mockImplementation(() => ({
        settings: Object.assign({}, mockSystemSettingsDefault.settings, {
            allowVisFullscreen: false,
        }),
    }))

    const { getByRole, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...defaultProps} />
        </WindowDimensionsProvider>
    )
    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View fullscreen')).toBeNull()
    })
})

test('does not render "Show interpretations" option if settings do not allow', async () => {
    useSystemSettings.mockImplementation(() => ({
        settings: Object.assign({}, mockSystemSettingsDefault.settings, {
            allowVisShowInterpretations: false,
        }),
    }))

    const { getByRole, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...defaultProps} />
        </WindowDimensionsProvider>
    )
    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('Show details and interpretations')).toBeNull()
    })
})

test('renders null if all relevant settings are false', async () => {
    useSystemSettings.mockImplementation(() => ({
        settings: {
            allowVisOpenInApp: false,
            allowVisShowInterpretations: false,
            allowVisViewAs: false,
            allowVisFullscreen: false,
        },
    }))

    const { container } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...defaultProps} />
        </WindowDimensionsProvider>
    )

    expect(container).toMatchSnapshot()
})

test('renders correct options for BAR in small screen', async () => {
    global.innerWidth = 480
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)

    const { getByRole, queryByTestId, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...defaultProps} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View as Map')).toBeTruthy()
        expect(queryByText('View as Chart')).toBeNull()
        expect(queryByText('View as Table')).toBeTruthy()
        expect(queryByTestId('divider')).toBeTruthy()
        expect(queryByText('Open in Data Visualizer app')).toBeNull()
        expect(queryByText('Show details and interpretations')).toBeTruthy()
        expect(queryByText('View fullscreen')).toBeTruthy()
    })

    global.innerWidth = 800
})

test('renders correct options for PIE in small screen', async () => {
    global.innerWidth = 480
    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)

    const props = Object.assign({}, defaultProps, {
        visualization: {
            type: 'PIE',
        },
    })

    const { getByRole, queryByTestId, queryByText } = render(
        <WindowDimensionsProvider>
            <ItemContextMenu {...props} />
        </WindowDimensionsProvider>
    )

    fireEvent.click(getByRole('button'))

    await waitFor(() => {
        expect(queryByText('View as Map')).toBeNull()
        expect(queryByText('View as Chart')).toBeNull()
        expect(queryByText('View as Table')).toBeNull()
        expect(queryByTestId('divider')).toBeNull()
        expect(queryByText('Open in Data Visualizer app')).toBeNull()
        expect(queryByText('Show details and interpretations')).toBeTruthy()
        expect(queryByText('View fullscreen')).toBeTruthy()
    })

    global.innerWidth = 800
})
