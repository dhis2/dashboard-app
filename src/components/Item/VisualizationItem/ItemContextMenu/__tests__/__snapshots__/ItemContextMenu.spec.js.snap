// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders just the button when menu closed 1`] = `
<div>
  <div>
    <button
      class="jsx-2371629422 secondary small icon-only"
      data-test="dashboarditem-menu-button"
      type="button"
    >
      <span
        class="jsx-2371629422 button-icon"
      >
        <svg
          color="#4a5768"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 10a2 2 0 110 4 2 2 0 010-4zm7 0a2 2 0 110 4 2 2 0 010-4zm7 0a2 2 0 110 4 2 2 0 010-4z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`renders null if all relevant settings are false 1`] = `<div />`;
